### ISSUE TITLE
[OWLAPI v. 4] *VisitorEx documented to return a NonNull value

### ISSUE THREAD

**Issue by ykazakov** [owlcs/owlapi#563]
Can I ask why in the OWLAPI 4 visitor interfaces (e.g., [OWLClassExpressionVisitorEx.java](https://github.com/owlcs/owlapi/blob/version4/api/src/main/java/org/semanticweb/owlapi/model/OWLClassExpressionVisitorEx.java)) the visitor methods should return non-null values? They are all annotated with `@NonNull`. If I implement a function for class expressions using a visitor, why it should necessarily return non-null values?
B.t.w., why `@NonNull` annotations are missing in the version5 branch?

**Comment by ignazio1977** [comment]
On 6 Oct 2016 3:53 p.m., "Yevgeny Kazakov" <notifications@github.com> wrote:
>
> Can I ask why in the OWLAPI 4 visitor interfaces (e.g.,
OWLClassExpressionVisitorEx.java) the visitor methods should return
non-null values? They are all annotated with @NonNull. If I implement a
function for class expressions using a visitor, why it should necessarily
return non-null values?

@Nonnull annotations are documentation annotations, not enforcement - it's
not necessary, just suggested.

The implementations used internally almost universally need to use the
returned value, and adding the annotations helped spot a few cases where
incorrect assumptions were made.

That said, if I had to design these interfaces now, I'd use Optional<T>
instead.

> B.t.w., why @NonNull annotations are missing in the version5 branch?
>

There is a @ParametersAreNotNullByDefault annotation on the package. It has
the same effect with less clutter.

Cheers,
I.

**Comment by ykazakov** [comment]
> @Nonnull annotations are documentation annotations, not enforcement - it's
not necessary, just suggested.

Well, usually there is a good reason to follow the documentation: if in java it is documented that a certain method should not accept null values, then it is better not to use it with nulls, or otherwise correctness is not guaranteed. Anyway, the reason why I am asking is that we use the [FindBugs](http://findbugs.sourceforge.net) static analyser, and it complains if the implementation of the visitors returns null values. I think Intellij have similar checkers.

> That said, if I had to design these interfaces now, I'd use Optional<T>
instead.

What would be achieved by that? If something is expected to be non-null, why should it be optional?

> There is a @ParametersAreNotNullByDefault annotation on the package. It has
the same effect with less clutter.

I see.
Do you, perhaps, mean @ParametersAreNonnullByDefault? (Google does not return much about @ParametersAreNotNullByDefault)
Are you sure it applies to the return values? There are [some doubts](http://stackoverflow.com/questions/7658353/is-parametersarenonnullbydefault-applies-to-method-return-values-too) about that.
Also, why the same would not work in owlapi 4?

**Comment by ignazio1977** [comment]
> we use the FindBugs static analyser, and it complains if the implementation of the visitors returns null values

Indeed, me too. I found that those places where a null was returned were more often bad code than good code. On balance, I thought we were better with then annotations than without.

About Optional, the advantage is that the possibility of a missing value is explicit, and with Java 8 writing code that is only executed on present values is quite compact.

> @ParametersAreNotNullByDefault

@ParametersAreNonnullByDefault indeed - apologies, I typed my previous answer on my phone and memory misled me.

The static analysers I've used interpret it as applying to return values as well - warnings on passing return values into non null arguments disappear with the annotation, and are restored if a method return is annotated with @Nullable.

The same change would work equivalently in OWLAPI 4, but I did not find out about the package annotation until after I had annotated most of the methods and fields. I didn't fancy reverting all the changes at the time...

**Comment by ykazakov** [comment]
OK, I see, this annotation is not really a documentation but a hint for the compiler to look for places when the (possibly unnecessary) null check for the returned value is made?
Hm... do you really have many of such unnecessary null checks?

> About Optional, the advantage is that the possibility of a missing value is explicit, and with Java 8 writing code that is only executed on present values is quite compact.

I don't understand. If the output is declared as optional, what prevents me from returning null?
Also, with optional values, does it mean that the output of every method should be explicitly checked to be absent? How does it differ from always checking the returned values for null? If, as you say, in most cases the return values are not null, then this seems to be wasteful.

**Comment by ykazakov** [comment]
OK, forget about the first point. Of course, if one declares arguments to be non-null then one should also declare the return values of other functions to be non-null to use them in such arguments.

With visitors returning null values I agree it is a rare situation, and in such cases a workaround is to add a field in such visitors to store the return (possibly null) value. That's what we do now to avoid the warnings.

**Comment by ignazio1977** [comment]
Not just unnecessary null checks, but (more importantly) missed necessary checks, warnings on values that are possibly null.

**Comment by ykazakov** [comment]
I just did a few tests out of curiosity.
I defined two interfaces:

```
public interface NonNullProducer {
	@Nonnull
	Object get();
}
```
```
public interface NonNullConsumer {
	void put(@Nonnull Object object);
}
```

and a test function:
```
void test(NonNullProducer producer, NonNullConsumer consumer) {
	Object o = producer.get();
	consumer.put(o);
}
```
Unsurprisingly, FindBugs did not find any problems here.
But neither in this piece of code with unnecessary null check:
```
void test(NonNullProducer producer, NonNullConsumer consumer) {
	Object o = producer.get();
	if (o != null) {
		consumer.put(o);
	}
}
```
Likewise, no warnings if I remove the @Nonnull annotation from the returned value of NonNullProducer and just call `consumer.put(producer.get())`
 
But the following, expectedly, gives me a warning "Null passed for non-null parameter", so FindBugs is clearly doing something:
```
void test(NonNullProducer producer, NonNullConsumer consumer) {
	Object o = null;
	consumer.put(o);
}
```
So my conclusion is that the @Nonnull annotations *on the returned values* do not do much in terms of detecting bugs.

Also, I think it would be surprising if the analyser would give me warnings when the result of the function `get()` without annotations is passed as the argument of `put(@Nonnull Object o)` with the annotation. This would mean that the absence of @Nonnull is interpreted as @Nullable, which I do not think should be the case as then there would be no reason to have @Nullable.
